### PR TITLE
LibJS/Bytecode: Implement async generators

### PR DESCRIPTION
**Pull request by Lubrsi**
Bytecode diff:
```
Summary:
    Diff Tests:
        +3960 ✅ (8.018%) +32 ❌ (0.064%) -3992 📝 (8.082%)
```
Tests passing: 47122/49388 (95.4%)

AST diff:
```
Summary:
    Diff Tests:
        +3964 ✅ (8.026%) +27 ❌ (0.056%) +1 💥️ (0.002%) -3992 📝 (8.082%)
```
Tests passing: 47322/49388 (95.8%)
        
https://www.youtube.com/watch?v=_ovdm2yX4MA

### CODE REVIEW

**Comment by trflynn89** [comment]
Couple of high level questions without getting into the diff:
Is this really a 1 commit change?
test-js tests?

File: Userland/Libraries/LibJS/Bytecode/ASTCodegen.cpp [hunk 1]
```diff
@@ -2404,32 +2479,17 @@ Bytecode::CodeGenerationErrorOr<void> ThisExpression::generate_bytecode(Bytecode
     return {};
 }
 
-static void generate_await(Bytecode::Generator& generator)
+static void generate_await(Bytecode::Generator& generator, Bytecode::Register received_completion_register, Bytecode::Register received_completion_type_register, Bytecode::Register received_completion_value_register, Bytecode::IdentifierTableIndex type_identifier, Bytecode::IdentifierTableIndex value_identifier)
 {
     VERIFY(generator.is_in_async_function());
 
     // Transform `await expr` to `yield expr`, see AsyncFunctionDriverWrapper
     // For that we just need to copy most of the code from YieldExpression
```
Comment (alimpfard) [inline Userland/Libraries/LibJS/Bytecode/ASTCodegen.cpp:2487]: This no longer seems to be happening.

**Review by alimpfard — COMMENTED** [review]
bytecode changes mostly LGTM, just one comment about a comment:

**Comment by trflynn89** [comment]
> Any idea what the new crashes are?

Seems like a spec bug to me. In bytecode mode, the crashes are:
```
test/built-ins/AsyncGeneratorPrototype/return/return-state-completed-broken-promise.js 📝 -> 💥
test/built-ins/AsyncGeneratorPrototype/return/return-suspendedStart-broken-promise.js  📝 -> 💥
```

These are asserting on infallible invocations to `AsyncGeneratorAwaitReturn`:
```c++
// ii. Perform ! AsyncGeneratorAwaitReturn(generator).
MUST(await_return());
```

However this operation is trivially fallible in step 6:
```c++
// 6. Let promise be ? PromiseResolve(%Promise%, completion.[[Value]]).
auto* promise = TRY(promise_resolve(vm, realm.intrinsics().promise_constructor(), completion.value().value()));
```

The current form of this AO was written [here](https://github.com/tc39/ecma262/commit/f4694412639e19acab2c0dd5baac247aa7d7a889). Before that, abrupt completions from `PromiseResolve` were specifically handled to reject the promise and prevent further throwing.

**Comment by trflynn89** [comment]
Ah there's a spec PR open for this already: https://github.com/tc39/ecma262/pull/2683

**Comment by Lubrsi** [comment]
All comments addressed except the 1 AST crash, which is:
```
💥️ /home/lukew/libjs-test262/test262/test/built-ins/AsyncGeneratorPrototype/next/request-queue-await-order.js                                                                                                                                                                     
                                                                                                                                                                                                                                                                                  
/home/lukew/Desktop/serenity-project/serenity/Userland/Libraries/LibJS/Runtime/Completion.cpp:117: JS::ThrowCompletionOr<JS::Value> JS::await(VM&, Value): Assertion `success.has_value()' failed.
```

Though I don't know if it's worth fixing this given that AST's `await` is very ad-hoc and AST is on it's way out.

**Review by linusg — APPROVED** [review]

**Pull request merged**